### PR TITLE
[Application Signals] Improve metric schema for EKS, Native K8s, EC2

### PR DESCRIPTION
**Pull request by bjrara**
# Description of the issue
Merge Application Signals changes to main.

# Description of changes
Changes are described in umbrella PR: https://github.com/aws/amazon-cloudwatch-agent/pull/1133.

Before the PR is merged, please make sure:
1. https://github.com/aws/amazon-cloudwatch-agent/pull/1180 is merged into application_signals_ga.
2. https://github.com/amazon-contributing/opentelemetry-collector-contrib/pull/197 is merged, and referenced in CWA.

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
_Describe what tests you have done._

# Requirements
_Before commit the code, please do the following steps._
1. Run `make fmt` and `make fmt-sh`
3. Run `make lint`

### CODE REVIEW

File: plugins/processors/awsapplicationsignals/internal/normalizer/attributesnormalizer_test.go [hunk 1]
```diff
@@ -8,8 +8,11 @@ import (
 
 	"github.com/stretchr/testify/assert"
 	"go.opentelemetry.io/collector/pdata/pcommon"
-	conventions "go.opentelemetry.io/collector/semconv/v1.6.1"
+	deprecatedsemconv "go.opentelemetry.io/collector/semconv/v1.18.0"
+	conventions "go.opentelemetry.io/collector/semconv/v1.22.0"
```
Comment (lisguo) [inline plugins/processors/awsapplicationsignals/internal/normalizer/attributesnormalizer_test.go:12]: similar to previous PR, rather just name this "semconv"

File: plugins/processors/awsapplicationsignals/internal/resolver/attributesresolver.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+func NewAttributesResolver(resolvers []appsignalsconfig.Resolver, logger *zap.Logger) *attributesResolver {
+	subResolvers := []subResolver{}
+	for _, resolver := range resolvers {
+		switch resolver.Platform {
+		case appsignalsconfig.PlatformEKS, appsignalsconfig.PlatformK8s:
+			subResolvers = append(subResolvers, getKubernetesResolver(resolver.Platform, resolver.Name, logger), newKubernetesResourceAttributesResolver(resolver.Platform, resolver.Name))
+		case appsignalsconfig.PlatformEC2:
+			subResolvers = append(subResolvers, newResourceAttributesResolver(resolver.Platform, AttributePlatformEC2, DefaultInheritedAttributes))
+		default:
+			if ecsutil.GetECSUtilSingleton().IsECS() {
```
Comment (lisguo) [inline plugins/processors/awsapplicationsignals/internal/resolver/attributesresolver.go:63]: Note: This is the main change. I am cross referencing this file with https://github.com/aws/amazon-cloudwatch-agent/blob/main/plugins/processors/awsappsignals/internal/resolver/attributesresolver.go

File: plugins/processors/awsapplicationsignals/internal/resolver/attributesresolver.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+func (h *resourceAttributesResolver) Process(attributes, resourceAttributes pcommon.Map) error {
+	for attrKey, mappingKey := range h.attributeMap {
+		if val, ok := resourceAttributes.Get(attrKey); ok {
+			attributes.PutStr(mappingKey, val.Str())
+		}
+	}
+	if _, ok := attributes.Get(attr.AWSLocalEnvironment); !ok {
+		if val, found := resourceAttributes.Get(attr.AWSHostedInEnvironment); found {
+			attributes.PutStr(attr.AWSLocalEnvironment, val.Str())
+		} else {
```
Comment (lisguo) [inline plugins/processors/awsapplicationsignals/internal/resolver/attributesresolver.go:115]: nit: I can see this part of the code getting large, we could have a helper func `GetLocalEnvironment`

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go [hunk 4]
```diff
@@ -607,16 +606,22 @@ func (e *kubernetesResolver) Process(attributes, resourceAttributes pcommon.Map)
...
 				e.logger.Debug("failed to Process ip", zap.String("ip", ipStr), zap.Error(err))
 				attributes.PutStr(attr.AWSRemoteService, "UnknownRemoteService")
 			}
 		}
 	}
 
+	if _, ok := attributes.Get(attr.AWSRemoteEnvironment); !ok {
+		if namespace != "" {
+			attributes.PutStr(attr.AWSRemoteEnvironment, fmt.Sprintf("%s:%s/%s", e.platformCode, e.clusterName, namespace))
+		}
```
Comment (lisguo) [inline plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:622]: is there a posibility of the AWSRemoteEnvironment not existing and namespace being empty? What should the AWSRemoteEnvironment be then?

File: translator/translate/otel/exporter/awsemf/translator.go [hunk 5]
```diff
@@ -142,14 +140,27 @@ func (t *translator) Translate(c *confmap.Conf) (component.Config, error) {
...
+	}
+
+	switch mode {
+	case config.ModeEKS:
 		return appSignalsConfigEks
-	} else if kubernetesMode == config.ModeK8sEC2 || kubernetesMode == config.ModeK8sOnPrem {
+	case config.ModeK8sEC2, config.ModeK8sOnPrem:
 		return appSignalsConfigK8s
+	case config.ModeEC2, config.ModeECS:
+		return appSignalsConfigGeneric
```
Comment (lisguo) [inline translator/translate/otel/exporter/awsemf/translator.go:160]: Should ModeEC2 have an EC2 config? Don't we have an EC2 resolver?

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go (hunk 4 above)
Comment (bjrara) [reply plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:622]: If the destination is outside the cluster, AWSRemoteEnvironment will be empty. It's an optional field.

File: translator/translate/otel/exporter/awsemf/translator.go (hunk 5 above)
Comment (bjrara) [reply translator/translate/otel/exporter/awsemf/translator.go:160]: The config template is actually unified regardless of the platform. The resolver is mainly for different attributes population.

File: plugins/processors/awsapplicationsignals/internal/normalizer/attributesnormalizer_test.go (hunk 1 above)
Comment (bjrara) [reply plugins/processors/awsapplicationsignals/internal/normalizer/attributesnormalizer_test.go]: Updated

File: plugins/processors/awsapplicationsignals/internal/resolver/attributesresolver.go (hunk 3 above)
Comment (bjrara) [reply plugins/processors/awsapplicationsignals/internal/resolver/attributesresolver.go]: Refactored.

**Review by lisguo — APPROVED** [review]

**Comment by lisguo** [comment]
Let's change the PR title to something like "[Application SIgnals] Improve metric schema for EKS, Native K8s, EC2".

**Review by lisguo — APPROVED** [review]

File: plugins/processors/awsapplicationsignals/internal/normalizer/attributesnormalizer_test.go [hunk 6]
```diff
@@ -106,18 +109,44 @@ func TestCopyResourceAttributesToAttributes(t *testing.T) {
 	}
 }
 
+func TestTruncateAttributes(t *testing.T) {
+	attributes := pcommon.NewMap()
+
+	longValue := make([]byte, 300)
+	for i := 0; i < 300; i++ {
+		longValue[i] = 'a'
+	}
```
Comment (jefchien) [inline plugins/processors/awsapplicationsignals/internal/normalizer/attributesnormalizer_test.go:118]: There's a https://pkg.go.dev/bytes#Repeat function if you wanted to use it.

**Review by jefchien — APPROVED** [review]

**Pull request merged**